CrazyCanvas/Source/Match/MatchServer.cpp

```diff
@@ -319,9 +319,30 @@ void MatchServer::FixedTickInternal(LambdaEngine::Timestamp deltaTime)
 			LOG_INFO("SERVER: Player=%u DIED", playerEntity);
 			DoKillPlayer(playerEntity);
 		}
-
 		m_PlayersToKill.Clear();
 	}
+
+	{
+		std::scoped_lock<SpinLock> lock2(m_PlayersToRespawnLock);
+		TArray<PlayerRespawnTimers> deferredPlayersToRespawn;
+
+		for (PlayerRespawnTimers& player : m_PlayersToRespawn)
+		{
+			player.second -= float32(deltaTime.AsSeconds());
+
+			if (player.second <= 0.0f)
+			{
+				deferredPlayersToRespawn.PushBack(player);
+			}
+		}
+
+		for (PlayerRespawnTimers& player : deferredPlayersToRespawn)
+		{
+			RespawnPlayer(player.first);
+			LOG_INFO("SERVER: Player=%u RESPAWNED", player.first);
+			m_PlayersToRespawn.Erase(m_PlayersToRespawn.Begin());
```

Comment:
Have you tested this code for multiple players? The erase function will always delete the first player even if it has not respawned yet. Consider erasing the player that respawned. The m_PlayersToRespawn could be a TSet to make it easier to use the erase function. Then you can do m_PlayersToRespawn.erase(player); (Asumes the std::pair has overloaded the required operators)

Reply:
Yes, I have tested this with 4 clients and spamming the built in "kill". Because the player who died first will always be the first item in the vector I assumed this was alright to do as doc says "This effectively reduces the container size by the number of elements removed, which are destroyed."

Reply:
Well, the code does not do what you think it does. You are always erasing the first player in the vector regardless if that player's timer has triggered. If you have 2 entities in the m_PlayersToRepawn and the second one is marked for respawn. Then you will erase the first one since you use m_PlayersToRepawn.Erase(m_PlayersToRepawn.Begin()); You should erase the entity that was triggered by the timer

Reply:
Are you sure? this will never be the case though, the case that an element will trigger before an element with a higher index. 

The full case is like follows: player dies -> added to TArray m_PlayersToRespawn -> if 5 sec pass -> add this element to deferredPlayersToRespawn -> if this TArray has any elements respawn those -> remove the first element(Because this will always be the player with lowest or equal timer left) in m_PlayersToRepawn.

Reply:
Yes, if you plan to always use the same time for every player it will work. However then a deferred erase is an unnecessary overhead. Then you should use erase in the first loop and skip the second one